lang/shared/src/main/scala/com/wavesplatform/lang/v1/compiler/compaction/ContractScriptCompactorV2.scala

```diff
@@ -0,0 +1,68 @@
+package com.wavesplatform.lang.v1.compiler.compaction
+
+import com.wavesplatform.lang.contract.DApp
+import com.wavesplatform.lang.v1.compiler.compaction.ContractScriptCompactor._
+
+import scala.annotation.tailrec
+
+object ContractScriptCompactorV2 extends ContractScriptCompactor[Vector[String]] {
+
+  def compact(dApp: DApp): DApp =
+    compact(dApp, State(0, Vector.empty), saveMeta = true)
+
+  def decompact(dApp: DApp): DApp = {
+    if (dApp.meta.originalNames.nonEmpty) {
+      compact(dApp, State(0, dApp.meta.originalNames.toVector), saveMeta = false)
+    } else {
+      dApp
+    }
+  }
+
+  @tailrec
+  protected def createCompName(oldName: String, state: State[Vector[String]], dApp: DApp, saveMeta: Boolean): CompactionResult[String, Vector[String]] = {
+    if (saveMeta) {
+      val compNameIdx = state.originalNames.indexOf(oldName)
```

Comment:
indexOf is O(N) operation, so it's better to form a temporary hash map and operate with it.